examples/schedulers/static_thread_pool.hpp

```diff
@@ -88,11 +89,275 @@ namespace example {
+
+                if (is_last_thread) {
+                  if (self.exception_) {
+                    std::execution::set_error((Receiver&&)self.receiver_, self.exception_);
+                  } else {
+                    std::visit([&](auto&& tupl) noexcept -> void {
+                      std::apply([&](auto&&... args) noexcept -> void {
+                        std::execution::set_value((Receiver&&)self.receiver_, std::move(args)...);
+                      }, std::move(tupl));
+                    }, std::move(self.data_));
+                  }
+                }
+              } else {
+                std::visit([&](auto& tupl) noexcept -> void {
+                  std::apply([&](auto&... args) noexcept -> void {
+                    auto [begin, end] = even_share(self.shape_, tid, total_threads);
+                    for (Shape i = begin; i < end; ++i) {
+                      self.fn_(i, args...);
+                    }
+                  }, tupl);
+                }, self.data_);
+
+                const bool is_last_thread = self.finished_threads_.fetch_add(1) == (total_threads - 1);
+
+                if (is_last_thread) {
+                  std::visit([&](auto&& tupl) noexcept -> void {
+                    std::apply([&](auto&&... args) noexcept -> void {
+                      std::execution::set_value((Receiver&&)self.receiver_, std::move(args)...);
+                    }, std::move(tupl));
+                  }, std::move(self.data_));
+                }
+              }
+            };
+          }
+        };
+
+      template <class SenderId, class ReceiverId, class Shape, class Fn, bool MayThrow>
+        struct bulk_receiver {
+          using Sender = std::__t<SenderId>;
+          using Receiver = std::__t<ReceiverId>;
+
+          using shared_state = bulk_shared_state<SenderId, ReceiverId, Shape, Fn, MayThrow>;
+
+          shared_state& shared_state_;
+
+          template <class... As>
+          friend void tag_invoke(std::execution::set_value_t, bulk_receiver&& self, As&&... as) noexcept {
+            shared_state& state = self.shared_state_;
+
+            if (state.shape_) {
+              if constexpr (MayThrow) {
+                try {
+                  using tuple_t = std::execution::__decayed_tuple<As...>;
+                  state.data_.template emplace<tuple_t>((As &&) as...);
+                  state.pool_.bulk_enqueue(&state, state.num_agents_required());
+                } catch (...) {
+                  std::execution::set_error((Receiver&&)state.receiver_, std::current_exception());
+                }
+              } else {
+                using tuple_t = std::execution::__decayed_tuple<As...>;
+                state.data_.template emplace<tuple_t>((As &&) as...);
+                state.pool_.bulk_enqueue(&state, state.num_agents_required());
+              }
+            } else {
+              std::execution::set_value((Receiver&&)state.receiver_, (As&&)as...);
```

Comment:
This is slightly off. In the `bulk_enqueue` case, what you'll be sending are xvalues referring to decay-copied values. Here, `As...` could contain lvalues. Once you fix up the completion_signatures as I suggest below, then you will not be declaring that you `bulk` might send lvalue. Instead, I think even in this code path, you need to decay-copy the values into the tuple and then move the elements into the `set_value` call.

Kinda sucks, I know, but them's the breaks.